interface/views.py

```diff
@@ -350,14 +349,58 @@ def code_view(request, pk, filename):
     submission = get_object_or_404(Submission, pk=pk)
     all_tas = submission.assignment.course.teaching_assistants.all()
 
-    if request.user not in all_tas:
+    if request.user not in all_tas and request.user is not submission.user:
         return HttpResponse(status=403)
 
     tree = tree_view(request, submission)
 
+    if filename == "code":
+        context = {
+            "tree": tree,
+            "pk": submission.pk,
+            "sub": submission,
+            "path": filename,
+        }
+        return render(request, "interface/code_view_homepage.html", context)
+
     file = extract_file(request, submission, filename)
 
-    context = {"file_content": file.read(), "tree": tree, "pk": submission.pk}
+    context = {
+        "file_content": file.getvalue(),
+        "tree": tree,
+        "pk": submission.pk,
+        "sub": submission,
+        "path": filename,
+        "file_exists": False,
+    }
+
+    if (
+        file.getvalue() != "The file is missing!"
+        and file.getvalue() != "The archive is missing!"
+    ):
+        table = table_maker(file)
+
+        comment = None
+        if request.method == "POST":
+            form = CommentForm(request.POST)
+            if form.is_valid():
+                comment = form.save(commit=False)
+                comment.submission = submission
+                comment.user = request.user
+                comment.path = filename
+                comment.line = request.POST.get("line", "")
+                comment.save()
```

Comment:
Maybe you can add hidden rows in your form so those can be filled in by form.save().